store/iavl/proof.go

```diff
@@ -0,0 +1,109 @@
+package iavl
+
+import (
+	"errors"
+	"fmt"
+
+	ics23 "github.com/confio/ics23/go"
+	"github.com/tendermint/tendermint/crypto/merkle"
+)
+
+const ProofOpIAVL = "iavlstore"
+
+type IAVLOp struct {
+	Key   []byte
+	Proof *ics23.CommitmentProof
+}
+
+var _ merkle.ProofOperator = IAVLOp{}
+
+func NewIAVLOp(key []byte, proof *ics23.CommitmentProof) IAVLOp {
+	return IAVLOp{
+		Key:   key,
+		Proof: proof,
+	}
+}
+
+func IAVLOpDecoder(pop merkle.ProofOp) (merkle.ProofOperator, error) {
+	if pop.Type != ProofOpIAVL {
+		return nil, errors.New(fmt.Sprintf("unexpected ProofOp.Type; got %v, want %v", pop.Type, ProofOpIAVL))
+	}
+	var op IAVLOp
+	proof := &ics23.CommitmentProof{}
+	err := proof.Unmarshal(pop.Data)
+	if err != nil {
+		return nil, err
+	}
+	op.Proof = proof
+
+	// Get Key from proof for now
+	if existProof, ok := op.Proof.Proof.(*ics23.CommitmentProof_Exist); !ok {
+		op.Key = existProof.Exist.Key
+	} else if nonexistProof, ok := op.Proof.Proof.(*ics23.CommitmentProof_Nonexist); !ok {
+		op.Key = nonexistProof.Nonexist.Key
+	} else {
+		return nil, errors.New("Proof type unsupported")
+	}
+	return op, nil
+}
+
+func (op IAVLOp) GetKey() []byte {
+	return op.Key
+}
+
+func (op IAVLOp) Run(args [][]byte) ([][]byte, error) {
+	switch len(args) {
```

Comment:
Both Existence and Nonexistence proofs are being handled by the same IAVLop here